### ISSUE TITLE
Support raw SQL queries without defining an entity type for the result

### ISSUE THREAD

**Issue by anpete** [dotnet/efcore#10753]
For Query Types it could be nice to not require a configuration call in OnModelCreating to add the type to the model. Instead, it may be possible to lazily add a type to the model on first access (usually query).

**Comment by mrlife** [comment]
@anpete I see where this allows raw sql to be run for a given context (existing class), but what about raw sql that is custom? Is there a way to just run a SQL command (e.g. a select) and get an object that contains the resulting rows (regardless of what tables the command includes)?

**Comment by anpete** [comment]
@mrlife A Query Type can be any shape you want. Define the type; add it to the model as a Query Type; and then you can query it using FromSql.

**Comment by mrlife** [comment]
@anpete That's definitely nice to have. I think the heart of what I'm looking for is a way to skip creating any model classes and just run a query and get a result. Are there any plans to support that, either in EF Core or in some way within a ASP.NET Core project?

**Comment by anpete** [comment]
@mrlife What would be the type of the result?

**Comment by mrlife** [comment]
@anpete This kind of query is a huge part of our business, where we use `join`s and `row_number` a lot.

I suppose there are many ways to approach how to dynamically create one or more objects to hold whatever is returned from the database (based on detected database column types). That's probably a long discussion on a desirable way to do that. It's not clear if there are any options currently available in ASP.NET Core that handle this.

Examples in .NET:
https://docs.microsoft.com/en-us/dotnet/api/system.data.sqlclient.sqldatareader
https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/ado-net-code-examples

**Comment by anpete** [comment]
@mrlife As you point out, ADO is the currently recommended way of doing this. NB: You can easily integrate ADO usage with EF, take a look at the [DbContext.Database](https://docs.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.dbcontext.database?view=efcore-2.0#Microsoft_EntityFrameworkCore_DbContext_Database) property.

It sounds like you want EF to do something here. Given that you don't want any mapping capabilities, what exactly would EF be adding over ADO? Perhaps providing an example code snippet would help.

**Comment by mrlife** [comment]
@anpete An ORM by itself would not provide any value with the types of queries we're running, however if EF Core with ADO is the way in ASP.NET Core to run any SQL and get a result set, then that's what we need.

I looked at [Configuring a DbContext](https://docs.microsoft.com/en-us/ef/core/miscellaneous/configuring-dbcontext), but it's not clear how to use it without an explicit model file. Is that possible?

**Comment by anpete** [comment]
@mrlife It is completely fine to just use ADO from ASP.NET Core - you don't need to use EF at all.

**Comment by Vasimovic** [comment]
Is this possible to do this now, can we add a model to the DbSet at runtime

`if (dbContext.Model.FindEntityType(typeof(T)) == null){
                DbQuery<T> query = dc.Query<T>();
}`

**Comment by Vasimovic** [comment]
RE: it may be possible to lazily add a type to the model on first access

That would be fantastic; although, I was wondering if you could also give us the option to do add a DbSet at runtime ourselves as well

Something like:

`if (dbContext.Model.FindEntityType(typeof(T)) == null) dbContext.AddQuery(typeof(T));`

**Comment by mrlife** [comment]
@anpete From your last comment, I'm having trouble finding documentation for this for aspnet core. Are you aware of any?

**Comment by smitpatel** [comment]
@mrlife - There is nothing special in ASP.NET core when it comes to using ADO.NET provider directly. We are not aware of any docs for that. But it wouldn't be much different from how you would use ADO.NET provider from any code.

**Comment by ajcvickers** [comment]
@Vasimovic What do you mean by, "implement this in EF 2.2"? Do you mean can you implement this in the EF code and send a PR?  Or do you mean, can you implement something on top of EF in your own code that will allow types to be added to the model dynamically?

**Comment by ajcvickers** [comment]
@Vasimovic The main limitation in EF Core to be aware of is that the model is read-only once it is being used by a context instance. (Mostly because multiple threads may be accessing it and mutation is not thread-safe.) So, the way to do this now is to build a new model when you need to add a new type, then create a new context instance that uses that model. Whether or not this is practical depends mostly on performance. That is, building a model is expensive, and caching many models will use a lot of memory.

**Comment by Vasimovic** [comment]
I need it in the same context, I am in the process of migrating a project from EF 6 to EF Core and have about 200 view-only models, adding the DbQuery<Model> to the context for 200 of them seemed a bit :smile: - Anyhow #10753 will be very useful for me

Thank you for taking your time to respond @ajcvickers much appreciated

**Comment by jenergm** [comment]
Hi @ajcvickers,
We would like to know if SqlQuery<TElement> will be released in the next version of .NET Core. It's so important to us to make compatible on changing our .NET Framework applications to .NET Core and to the Cloud in the close future.
Have you any schedule for this implementation?
Cheers,

**Comment by ajcvickers** [comment]
@jenergm The milestone for each issue indicates the release in which a fix for the issue is planned to ship. Currently we have tentative planning (subject to change) for two releases: 2.2, and 3.0. Issues, like this one, that are in the Backlog milestone are planned to be post 3.0.

**Comment by jenergm** [comment]
Hi @ajcvickers. Thanks for sharing the road map.

**Comment by jenergm** [comment]
Hi @ajcvickers , how's going?
I overhear .NET Core 3.0 will release starting 2019. https://blogs.msdn.microsoft.com/dotnet/2018/05/07/net-core-3-and-support-for-windows-desktop-applications/

Are you know if SqlQuery<TElement> will be released in that next version of .NET Core?

Best regards,

**Comment by ajcvickers** [comment]
@jenergm Not as currently planned.

**Comment by jenergm** [comment]
Hi @ajcvickers ,
You told us on this year 9 Aug that would be in the Backlog milestone planned to be post in 3.0 version as later.
There are some problem to don't put it on?
We are changing some important applications to .NET Core run in the cloud and this SqlQuery&lt;TElement&gt; is the key to success.
I wondering when you intend release, because our schedule is compromised.
Can you share us some dates about it? Will this be done?
Best regards,

**Comment by pantonis** [comment]
Any news on this one? When do you plan to release it?

**Comment by ajcvickers** [comment]
When looking at this, consider also the issues raised in #14063. Specifically how would different mappings--keys, facets, etc--can be handled for the same CLR type dynamically added in different ways.

**Comment by ericnewton76** [comment]
@ajcvickers I think the implied implementation for SqlQuery<Tresult> here is much simpler implementation than is being imagined from your team.

Not sure how EF6 does it, but just basic column name to Tresult property name mappings and type conversions handled via Convert.To[PropertyType] is requested here.  Eventually people will complain about the performance of it, but frankly it'll always theoretically be x^n faster than the actual network IO to grab the query from a real database connection.

A way to address performance would be mappings but not attached to the context but a separate mapping attached to the Tresult type itself (via hashset or whatever) with thread safe construction of the initial mapping.  Once thats baked in, that Tresult cant be fed query shape A and query shape B... it must be used always for query shape A and only if query shape B is a superset of A will that same mapping work... otherwise the developer must use a different Tresult for query shape B and theres no equivalency or whatever to Tresult-A.

If one were to attempt to implement this (like me) I would almost imagine the SqlQuery<Tresult> code to ironically be almost entirely separate code paths from all the model mapping caching etc code in the rest of EFcore.  Sometimes we just need a super simple column name to property name deserializer... lol

**Comment by divega** [comment]
In https://github.com/aspnet/EntityFrameworkCore/issues/1862#issuecomment-451671168, @davidbaxterbrowne described a SqlQuery extension method that works by creating an instance of a derived DbContext type that is generic on the new type to be mapped. And I actually think this is a promising approach:

- Creating a new DbContext type and instance implies some overhead, but this way it leverages existing building blocks, including model caching for the DbContext. 

- It is constrained compared to query types in the sense that you cannot customize the mapping, but I think a story that sets this constraint for ad-hoc types and points to mapping types explicitly in OnModelCreating for more advanced scenarios works well. Incidentally, this is the feedback from @ericnewton76 above. 

- The only wrinkle I can see is what happens if the ad-hoc type references (through navigation properties) types already explicitly mapped in the original DbContext. Should the model configuration for the closure of types referenced in this way be borrowed from the root DbContext?

**Comment by davidbaxterbrowne** [comment]
I just added an issue tangentially related to this, where constructor binding fails for anonymous types.

https://github.com/aspnet/EntityFrameworkCore/issues/14336

If we can dynamically introduce query types, then people will want to use Anonymous Types, as those (and Tuples) are the only kinds of types that can be declared inline.  So allowing the query result shape to be declared at in the same scope as the query.  

Tuples can't be used as they are not Reference types, and so can't be Entities or Query Types (possible enhancement), but Anonymous types should work fine.  They are generated as types with read-only properties and a constructor to set all of the property values.  So they _should_ work fine with Constructor binding.

**Comment by smitpatel** [comment]
> The only wrinkle I can see is what happens if the ad-hoc type references (through navigation properties) types already explicitly mapped in the original DbContext.

Do we need to support navigations on QueryTypes? What does it represent in terms of database?

**Comment by ericnewton76** [comment]
I dont see why navigations needs to be supported.  Again, I believe this is being way over engineered.

Sometimes I just need a query projection deserialized into POCO objects.  SqlQuery<T> in EF noncore does this.

I dont expect support for navigation, I dont even expect being able to pass whatever pops out and serialize to SQL strings... (how does one serialize a projection?)

I just need a simple way to take the columns of a resultset, map them to the like-named columns of a POCO class, and return it.

**Comment by AndriySvyryd** [comment]
Related to https://github.com/aspnet/EntityFrameworkCore/issues/2282

**Comment by John0King** [comment]
from #17516 :

I'm suggestion add a new method that create `IQuerable<T>` directly from sql,  and the property map must use Sql to match model  instead of use model to match sql so there no need to configurate the query model in the dbContext.

```C#
DbContext.FromSqlRaw<T>(string sql)
DbContext.FromSqlRaw<T>(string sql, T objectProjection> // some idea of dynamic/anonymous support , .FromSql("SELECT **", new { Name= "", Age = 0, BirthDay = (DateTime?)null })
```

**Comment by ErikEJ** [comment]
@john0king See https://github.com/aspnet/EntityFrameworkCore/issues/1862#issuecomment-451671168

**Comment by jez9999** [comment]
Another vote from me on implementing SqlQuery<T> in EF Core.  It was extremely useful.

**Comment by AndriySvyryd** [comment]
**Design proposal**: instead of creating an `IEntityType` use a [multi-column value converter](https://github.com/dotnet/efcore/issues/13947) as it's much more lightweight.

**Comment by angelaki** [comment]
Is this feature still planned for EFC6? Or does a preview already contain it?

**Comment by roji** [comment]
@angelaki this feature has been punted for EF Core 6.0 (as per the milestone above).

**Comment by angelaki** [comment]
Oh sorry, just seen it. Isn't it pretty trivial? Anyways, thank you for drawing attention.

**Comment by langdonx** [comment]
I've been using this, sourced from a few different places ([#1862](https://github.com/dotnet/efcore/issues/1862#issuecomment-451671168) and StackOverflow).  It works great, but is somewhat annoying because (1) you have provide a type on the left side of your assignment (can't use `var`) because of the `dynamic` and (2) you have to create a class (can't use an anonymous object).

I prefer this over others I've seen because I like dealing with parameters as an object vs an array.

Usage:

```
public class GetCustomerData
{
	public int Id { get; set; }
	public string Name { get; set;}
}

dynamic sqlParameters = new ExpandoObject();
sqlParameters.CustomerId = 1;
sqlParameters.Skip = 0;
sqlParameters.Take = 10;

var sql = @"
select CustomerId as Id, Name
  from Customer
 where customerid = @CustomerId
 order by CustomerId
offset @Skip rows fetch next @Take rows only";

List<GetCustomerData> rows = await context.SqlQueryAsync<GetCustomerData>(sql, sqlParameters);
```

Implementation:

    public static class SqlQueryPolyfill
    {
        // here because SqlQuery<T> doesn't exist in EF Core https://github.com/dotnet/efcore/issues/10753
        public static async Task<List<T>> SqlQueryAsync<T>(DatabaseFacade database, string sql, dynamic parameters = null) where T : new()
        {
            using (var command = database.GetDbConnection().CreateCommand())
            {
                command.CommandText = sql;
                database.OpenConnection();

                if (parameters != null)
                {
                    if (parameters is ExpandoObject)
                    {
                        foreach (var kvp in parameters)
                        {
                            var parameter = command.CreateParameter();
                            parameter.ParameterName = kvp.Key;
                            parameter.Value = kvp.Value;
                            command.Parameters.Add(parameter);
                        }
                    }
                    else
                    {
                        foreach (var prop in parameters.GetType().GetProperties(BindingFlags.Instance | BindingFlags.Public))
                        {
                            var parameter = command.CreateParameter();
                            parameter.ParameterName = prop.Name;
                            parameter.Value = prop.GetValue(parameters, null);
                            command.Parameters.Add(parameter);
                        }
                    }
                }

                using (var reader = await command.ExecuteReaderAsync())
                {
                    if (reader.HasRows)
                    {
                        var dt = new DataTable();
                        dt.Load(reader);
                        return DataTableToList<T>(dt);
                    }
                    else
                    {
                        return new List<T>();
                    }
                }
            }
        }

        private static List<T> DataTableToList<T>(DataTable table) where T : new()
        {
            var list = new List<T>();

            var typeProperties = typeof(T)
                .GetProperties()
                .Select(propertyInfo => new
                {
                    PropertyInfo = propertyInfo,
                    Type = Nullable.GetUnderlyingType(propertyInfo.PropertyType) ?? propertyInfo.PropertyType
                })
                .ToList();

            foreach (var row in table.Rows.Cast<DataRow>())
            {
                T obj = new T();

                foreach (var typeProperty in typeProperties)
                {
                    if (row.Table.Columns.Contains(typeProperty.PropertyInfo.Name))
                    {
                        object value = row[typeProperty.PropertyInfo.Name];
                        object safeValue = value == null || DBNull.Value.Equals(value)
                            ? null
                            : Convert.ChangeType(value, typeProperty.Type);

                        typeProperty.PropertyInfo.SetValue(obj, safeValue, null);
                    }
                }
                list.Add(obj);
            }
            return list;
        }
    }

**Comment by ErikEJ** [comment]
I have a NuGet package that implements SqlQueryAsync here: https://www.nuget.org/packages/ErikEJ.EntityFrameworkCore.SqlServer.SqlQuery/

**Comment by RainingNight** [comment]
Any news?

**Comment by roji** [comment]
@RainingNight this issue is currently on the backlog, so is not planned for the EF Core 6.0 release.

**Comment by angelaki** [comment]
@RainingNight I too hoped for this feature quite long until I found https://github.com/DapperLib/Dapper. I use EFCore for my regular / more complex queries now and Dapper for primitive ones. And they both together work like a charm.

**Comment by ErikEJ** [comment]
@RainingNight  and @angelaki You can also use [this](https://github.com/dotnet/efcore/issues/10753#issuecomment-849615018)

**Comment by CwjXFH** [comment]
Yea, sometimes we need to execute the complex query by calling the RawSql method, but the method only returns the Entity types. So I have to define extension methods for EFCore use Dapper.

**Comment by ErikEJ** [comment]
@stephajn yes it will

**Comment by AraHaan** [comment]
I would love if this feature would also allow for raw queries that:
do not return things from ``SELECT`` (like ones that set the actual outputs to an parameter (stored procedure)), this would actually remove a lot of my ADO.NET Code as then I would know that:
- my fromCode / fromData converters for specific types will run (with ADO.NET there is no guarantee of that happening).

**Comment by ajcvickers** [comment]
Unfortunately, this feature is dependent on #13947, which is being cut from the 7.0 release. We know people are waiting on this, and it will be a high priority for EF8.

**Comment by smitpatel** [comment]
Few design notes.
- This will hare API with #11624 so will be composable.
- The type of `T` will be non-scalar type
- We will match the property on the type to column names in projection.
- If `T` turns out to be an entity type in model then we will throw exception. Users can use `DbSet<T>.FromSql` to achieve same result. This avoids the ambiguity if the configuration on entity type will be used or not (like custom column names etc) and if the resulting query will be tracked or not.

**Comment by stephajn** [comment]
> @jenergm That will be November 2023.

That's a real shame. For all of the time that EF Core has been out and this feature still isn't there, it makes me sad that it will be put off for yet another year.

**Comment by AraHaan** [comment]
> @jenergm That will be November 2023.

When will be EFCore 10?

**Comment by pantonis** [comment]
I think the problem is not the EF Core team (which they do fantastic job) but Microsoft itself.  They left the team with very few resources compared with the importance and size of EF Core. I don't know if we can do it but it would be good to open a petition somewhere and ask from Microsoft to allocate much more resources to the EF Core team. Important and very core features are missing or took years to develop wheras if the team had more resources, I strongly believe it they would have been implemented. I have seen numerous issues where people (incl my self) complaining all the time for the lack of core features. So I believe we should stop complaining and open a petition somewhere to help the EF Core team getting more resources.

**Comment by AraHaan** [comment]
> I think the problem is not the EF Core team (which they do fantastic job) but Microsoft itself. They left the team with very few resources compared with the importance and size of EF Core. I don't know if we can do it but it would be good to open a petition somewhere and ask from Microsoft to allocate much more resources to the EF Core team. Important and very core features are missing or took years to develop whereas if the team had more resources, I strongly believe it they would have been implemented. I have seen numerous issues where people (incl my self) complaining all the time for the lack of core features. So I believe we should stop complaining and open a petition somewhere to help the EF Core team in getting more resources.

Or better yet, work on EFCore as if we were team members but first learn the codebase.

**Comment by roji** [comment]
> When will be EFCore 10? Also why cant even releases be released on even years and not odd years (for LTS) and odd releases for odd years?

@AraHaan releases happen every year around November, with LTS (even version numbers) and non-LTS (odd version numbers) releases alternating. Since 7.0 is being released in 2022, that means that 10.0 will be released in 2025. I'm not sure why it's important for even version numbers to be released on even years - I definitely wouldn't change the whole schedule of .NET just to make that happen.

On a personal note from me... In my view, this feature really is a convenience that is outside EF's core functionality; Dapper was basically built to provide this exact functionality (executing raw SQL and materializing the results as arbitrary .NET POCOs), so you can very easily achieve this in your code today, without needing any special support from EF. I understand that some people want a single package (EF) to take care of all their data needs (or want to use Microsoft-only software), but I personally believe that's not a very sustainable approach in today's software world.

Note that other people in the EF team (possibly most of them) don't agree with me on this point, and we definitely are planning to implement this. But the fact that an alternative route/workaround (Dapper) exists makes this less urgent than some other features being implemented, at least in my view.

**Comment by ErikEJ** [comment]
I for one disagree on the priority of this and think having a single Microsoft supported solution is important.

But in the meantime I have a Nuget package that enables this scenario: https://www.nuget.org/packages/ErikEJ.EntityFrameworkCore.SqlServer.SqlQuery/6.0.0-preview1

**Comment by angelaki** [comment]
> I for one disagree on the priority of this and think having a single Microsoft supported solution is important.
> 
> But in the meantime I have a Nuget package that enables this scenario: https://www.nuget.org/packages/ErikEJ.EntityFrameworkCore.SqlServer.SqlQuery/6.0.0-preview1

As @roji already mentioned, this feels like reinvent the wheel. Sure, if EF offered it, I'd use it. But Dapper is a quite perfect solution for this usecase, worth every of its 200kb ;). So I think this feature can actually be put far in the backlog.

**Comment by ErikEJ** [comment]
I think doing this is very much in line with  one of the main principles of EF Core:

Eliminates the need for most of the data-access code that typically needs to be written.

**Comment by AraHaan** [comment]
Agreed, I think having official code that seems minor to some could be major to users who actually need this trivial feature without them needing to depend on yet another package (that could come with it's own security problems). With Microsoft maintained code, there is an guarantee that if an security issue is ever found, they will patch it right away depending on how significant it may be which for me I find is a much needed bonus over 3rd party libraries that may or may not be maintained anymore and so security issues would go unpatched with no way to patch it (other than forking and maintaining it yourself).

**Comment by roji** [comment]
@ErikEJ 

> Eliminates the need for most of the data-access code that typically needs to be written

Interesting coming from the maintainer of the EF Power Tools :wink: Though I'm being somewhat facetious, of course (a UI isn't code). The point is more that doing raw SQL via Dapper is almost no extra code compared to what EF would provide via this issue (a simple extension method over a context could easily implement this).

@AraHaan

>  I think having official code that seems minor to some could be major to users who actually need this trivial feature without them needing to depend on yet another package [...] With Microsoft maintained code, there is an guarantee [...]

I very strongly believe that in order to thrive, the .NET ecosystem needs to grow beyond this mindset; no other ecosystem out there has a single company such as Microsoft providing so much of its software, and companies/users insistent on using only software from that company. The idea that software is only official, can only be secure or only gets properly maintained if it comes from Microsoft is IMHO problematic and a throwback to the pre-OSS days. Expecting Microsoft to cover any and all software needs out there is unsustainable and unrealistic; people simply have to get used to software coming from other sources, just like the rest of the software world works.

**Comment by jez9999** [comment]
> I very strongly believe that in order to thrive, the .NET ecosystem needs to grow beyond this mindset; no other ecosystem out there has a single company such as Microsoft providing so much of its software, and companies/users insistent on using only software from that company. The idea that software is only official, can only be secure or only gets properly maintained if it comes from Microsoft is IMHO problematic and a throwback to the pre-OSS days. Expecting Microsoft to cover any and all software needs out there is unsustainable and unrealistic; people simply have to get used to software coming from other sources, just like the rest of the software world works.

I don't think this is a given at all.  When I dabbled in using JS on the server-side one of the things that overwhelmed me and put me off frankly was that for virtually any task, there were 1000 different libraries offering the functionality and aside from a Google search for general popularity, which tended to leave 5 or 10 main competitors, there was very little way to choose between them.  When you did pick one, you'd have trouble getting support if there was a problem because everything was so balkanized.  There's something to be said for lots of useful functionality being provided by a central provider that most people tend to use.

**Comment by roji** [comment]
@jez9999 yes, in a mature ecosystem there are various alternatives to choose from, and a good ranking/usage system is key to helping users making the right decision; choice is a good thing. Saying that one company should provide everything isn't a viable or desirable alternative, IMHO, and stifles ecosystem progress to whatever that company happens to be prioritizing at the moment because of its considerations.

**Comment by vigouredelaruse** [comment]
tools like power apps and azure data factory that connect logic and data, and no (db layer) implementation details are necessarily surfaced to say a power user, and we all want to duplicate that functionality, me specifically for a cms framework i'm building using ef core but 
* efcore does not claim to support workflows that require 'magical' materialzation of arbitrary models as db schema at runtime, without statically compiled assets being involved
* efcore does not claim to support 'magical' materialization of query results against arbitrary table and column sets

i see that this issue is a step in the direction of unlocking the use cases hinted at here and so i'm adding some more justification and reasons for customer interest

regards

**Comment by roji** [comment]
@vigouredelaruse raw queries for arbitrary CLR types is something you can already easily do with Dapper; if that's the main thing your application needs to do, then there's really little reason to use EF.

**Comment by Mardoxx** [comment]
That’s not the point. Many things could be done with external libraries.
These community maintained projects are a lot more risky to use from
functionality provided in core framework.

**Comment by roji** [comment]
@Mardoxx EF Core cannot be a universal answer for every need out there, and treating community projects as inherently risky is IMHO one of the major problems in the .NET ecosystem; other ecosystems (Java, Python) depend on community projects and are doing just fine, whereas in the .NET ecosystem many users are still fixated on using only Microsoft software. I can point to various .NET community projects which are doing better than their Microsoft counterpart, so I'd recommend evaluating which projects you're using based criteria other than just Microsoft or not.

**Comment by vigouredelaruse** [comment]
@rojj >  raw queries for arbitrary CLR types is something you can already easily do with Dapper; 

true but my inquiries on this are related to scenarios where the 'end user developer' either does not, should not or must not have access to the db layer.  so we're discussing how to build something that enables low code development workflows, client side db query definition and the like
 
i'm building such a thing shaped as a multitenant headless cms framework satisfying at least
* disruption of the ability to inject dbcontexts of other tenants running on the instance (hoped to be accomplished here by removing the need for 'end user developers' to write then)
* support for easy *post-deployment* schema definition/extension and query execution (partially resolved by using odata as i mentioned) 
* unlock (not just) client side db query without framework attempting to forsee and provide (ship with releases) all queries users of the framework are ever going to want to use

as i mentioned for the scenario i'm concerned about odata unlocks the capability to give developers the ability to write their own queries, within a server side mediated query sandbox of course

unlocking the ability to define and deploy the schema at runtime is harder. 

i think the answer for this will be to leverage json because  json schema definition can be decoupled from the schema's db provider materialization

so the first cut attempt would be a 'bag of json' column containing arbitrary entities with a minimal common property set (id, timestamp, shard, blahblah) . like azure table storage.

i can feel this starting to get handwavy so over

**Comment by roji** [comment]
@vigouredelaruse I'm not really following... Regardless of the specific scenario you're trying to enable, the discussion here is about executing a SQL query and materializing the result as an arbitrary .NET type. This is something that's already well-supported by Dapper, and this issue tracks adding the same capability to EF.

I can see that making sense if your application generally works with EF since it uses statically-known types, but *occasionally* needs to do such a dynamic query. However, if your application mainly interacts with dynamic database data via raw SQL queries - as seems to be the case from your description above - then EF is probably not the right tool for the job.

**Comment by vigouredelaruse** [comment]
@roji noted this issue is about raw sql => clr mapping, well supported by dapper

and, the project i'm concerned about has already chosen efcore specifically to delegate schema management and sql code generation. i won't be dirtying my keyboard with sql but the project's users may want to, and i want to let them. 

the question is what mix of orm and client query mechanisms will deliver running code faster. 

i've found that mixing odata (for queries) and efcore (for sql and c# model codegen) is working quite well.

starting fom there please consider the following scenario
* i publish a .net core template and it's supporting nugets
*  someone uses the template in visual studio or dot net new to start a website project with builtin cms functionality,
* they spin up a website that allows users to create and query database entities and use those entities on their content collections *without recompiling the project* but perhaps by using a dto clr poco with a json payload

now that efcore supports json things get easier. store the custom entities as json and the model as jsonschema. 

now i need to hydrate arbitrary classes to refer to in controller action methods

i hope i've clarified the scenario enough to show how this issue is of interest. by itself it doesn't seem to fit with my scenario but in context i assure you it does

regards

**Comment by roji** [comment]
> they spin up a website that allows users to create and query database entities and use those entities on their content collections without recompiling the project but perhaps by using a 'well known' dto clr poco with a json payload
> [...]
> now i need to hydrate arbitrary classes to refer to in controller action methods (and document/operation authorization handlers).

This is specifically not a mode that's well-supported by EF. You can theoretically do completely dynamic model building and add arbitrary user POCO types to the model. But the real question is what querying even looks like in this kind of scenario; sure, you can dynamically construct LINQ expression trees which query your dynamic user POCOs, but there's very little advantage of using EF in this way over just using SQL.

> now that efcore supports json things get easier

Not really; EF's JSON support still requires a well-known, non-dynamic model for the stuff you put inside JSON documents - unless you want to just deal with JSON data as strings, and take handle serialization/deserialization yourself (that has always been possible, EF7 doesn't make that mode better).

To summarize, EF (like most other ORMs in the world) simply isn't meant to be used in scenarios where the schema and queries are completely dynamic. Regardless of that, .NET already has good ways of materializing arbitrary SQL results to CLR POCOs - that's exactly what Dapper does. We do intend to provide that ability in EF (that's what this issue is about), but if that's going to be your **main** use of EF, I'd advise just going with Dapper.

**Comment by vigouredelaruse** [comment]
ackowledged. apologiess for the distraction

**Comment by teliaz** [comment]
As a workaround you can create a empty view with the appropriate return types and use 
`FromSqlRaw` 

Works fine on ef6, ef7. 

Example follows above: 

One solution I got to maintain all dtos in the database was to create a view 
with a dummy return type. 

**The view** 
```sql
CREATE OR REPLACE VIEW public.result_function_metrics
AS SELECT 0::double precision AS date_year,
    0::double precision AS date_month,
    0::bigint AS metric,
```

**The function** 
```sql
CREATE OR REPLACE FUNCTION fx_reports_goals (
	_user_id int8, 
	_date_from timestamp with time zone, 
	_date_to timestamp with time ZONE
)
RETURNS SETOF result_function_metrics
AS 
$$
BEGIN

... 

END
$$
LANGUAGE plpgsql;
```



```c#
var query = @"SELECT * FROM fx_reports_goals(
            @pUserId,
            @pDateFrom,
            @pDateTo);";
      var parameters = new NpgsqlParameter[3];

parameters[0] = new NpgsqlParameter("pUserId", NpgsqlTypes.NpgsqlDbType.Bigint);
parameters[0].Value = filter.userID ?? (object)DBNull.Value;
--- 
// same for the rest of params
--- 

var results = await db.ResultFunctionMetric.FromSqlRaw(query, parameters).ToListAsync();
```

Notice that db contains the dbSet `ResultFunctionMetric` scaffolded from the empty view.

TL/DR
It is easier to set return type from the function or stored procedure add a dummy view with the expected return values and as a bonus you still get that everytime you scaffold (database first approach)

**Comment by ErikEJ** [comment]
@teliaz this is exactly what EF Core Power Tools does for SQL Server. Wonder if someone could help do the same for Postgres?

**Comment by ErikEJ** [comment]
Amazing. So this will be in daily build now?

**Comment by ajcvickers** [comment]
@ErikEJ Should be.

@jenergm As the milestone indicates, it will be released in EF Core 8. However, it should be available now in the [daily builds](https://github.com/dotnet/efcore/blob/main/docs/DailyBuilds.md) and will be included in preview 1 of EF8.

**Comment by goenning** [comment]
@ajcvickers just tried this feature today and I love it! This can easily replace Dapper for many and allow us to incrementally adopt EF Core.

One question though: will it support scalar values? This doesn't seem to work (only tested with npgsql).

```csharp
_db.Database.SqlQuery<string>($"SELECT id FROM environments WHERE key = {appKey}").FirstOrDefault();
```

**Comment by Misiu** [comment]
> will it support scalar values?

That would be awesome! Waiting for more info.

**Comment by ajcvickers** [comment]
@goenning @Misiu Scalar values were already supported in EF7, and should still work. If this isn't the case, then please open a new issue and attach a small, runnable project or post a small, runnable code listing that reproduces what you are seeing so that we can investigate.

**Comment by Misiu** [comment]
@ajcvickers I've created a new issue showing this error: https://github.com/dotnet/efcore/issues/30447 In my case I'm using SqlServer.

**Comment by goenning** [comment]
Thanks @Misiu thsts the same error message I get using Postgres

**Comment by borisdj** [comment]
@ajcvickers I have posted related Q on SO:
https://stackoverflow.com/questions/76331657/dynamic-sub-select-using-expression-with-fromsqlraw-in-net-core-v7-with-cast
Can it be linked here, and can anyone take a look, or should new issue be opened.

**Comment by ite-klass** [comment]
@borisdj The linked question is confusing overall, but it adds a class definition to the DB context - so no, it's not directly relevant to this ticket - which is specifically about non-entity result types

**Comment by pcapozzi** [comment]
would it be possible to show an example of how to call a stored procedure with this new functionality?

**Comment by ajcvickers** [comment]
@pcapozzi See https://learn.microsoft.com/en-us/ef/core/what-is-new/ef-core-8.0/whatsnew#raw-sql-queries-for-unmapped-types

**Comment by pcapozzi** [comment]
Thanks for that information.  It seems that EF8 will give some exciting new
potential.  My only comment is that when you mention in the docs  ' The
code shown here comes from RawSqlSample.cs.'

Although it does seem to be part of NewInEFCore8 project, I have never been
able to download  a specific project.  They all seem to be part of the
'EntityFramework.Docs' solution which is now up to what seems like 100
terabytes.  It would be very useful to be able to 'clone' individual
projects.

Thanks for your help

Regards,

PatC

**Comment by DavidThielen** [comment]
This is added so great. But want to add that this is essential for calls that use FreeTextTable() and ContainsTable() in SQL Server.
